stdenv: bootstrap riscv64-linux (Step 1)

oxalica opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change
Fix bootstrap packages and bootstrap `stdenv` on native `riscv64-linux` platform.

Related: #101651

**UPDATE  2021.11.27**

Rebased and fixed. Currently it doesn't contain bootstrap binary hashes. As suggested by @r-burns https://github.com/NixOS/nixpkgs/pull/115406#issuecomment-950109021, we should build it on Hydra for trusted binaries.

**UPDATE  2021.10.23**
Rebased and re-bootstrapped with [QEMU VM](https://github.com/NixOS/nixpkgs/issues/101651#issuecomment-852725823). Thanks @zhaofengli.

<details>
<summary>Previous attempt</summary>

Files ready to be uploaded to `tarballs.nixos.org`: (Currently on my cachix https://riscv64.cachix.org)
- /nix/store/yffnjz4sny3ygbm59wk69yk56ja0lyk8-busybox
  Nar hash: `sha256:d67bc5ef5c550bc1e2b86b27fe20b9ffadca42b9d104ae0f27957794be62ac5f`

- /nix/store/gaa783s4y4w14wyc94lh5pvbpyszbvhc-bootstrap-tools.tar.xz
  File hash: `sha256:541b51c0dcc4e9fb04652664ff03581c73c16e24198e4030c8ff205fc1114c3d`

###### Steps **I** did

1. Cross build bootstrap tools for `riscv64-linux`, with build platform `x86_64-linux`.
```shell
nix build -f pkgs/stdenv/linux/make-bootstrap-tools-cross.nix riscv64.build -o result-rv64-bootstrap-tools-stage0
```

2. Get hashes to fill `pkgs/stdenv/linux/bootstrap-files/riscv64.nix`.
   Then add fixed-output paths into store since our bootstrap file URLs are fake.
```shell
nix hash path --type sha256 --base16 ./result-rv64-bootstrap-tools-stage0/on-server/busybox
nix hash file --type sha256 --base16 ./result-rv64-bootstrap-tools-stage0/on-server/bootstrap-tools.tar.xz
# ... hashes to fill

nix store add-path ./result-rv64-bootstrap-tools-stage0/on-server/busybox
nix store add-file ./result-rv64-bootstrap-tools-stage0/on-server/bootstrap-tools.tar.xz
```

3. Setup a native `riscv64-linux` platform.
   ~For me, [HiFive-Unleashed](https://www.sifive.com/boards/hifive-unleashed) is used.
   Cross build `nix` for `riscv64-linux`, copy closures to the riscv64 platform, and setup `/nix` and `PATH` correctly.~
   I currently don't have any native machine, so I setup an QEMU VM for the bootstrap build.
   Note that some packages like `coreutils` must be built on native machine or QEMU VM, instead of binfmt (QEMU user mode), to pass all tests . See Notes 3 below.

4. Remote build native bootstrap tools on `riscv64-linux` with previous cross bootstrap tools.
```shell
nix build -f pkgs/stdenv/linux/make-bootstrap-tools.nix build --arg localSystem '{ system = "riscv64-linux"; }' -L --keep-going --builders 'ssh://rv64_machine_or_vm riscv64-linux - 4' -j0 -o result-rv64-bootstrap-tools-stage1
```

5. Update bootstrap file hashes to native ones in `pkgs/stdenv/linux/bootstrap-files/riscv64.nix`.
```shell
nix hash path --type sha256 --base16 ./result-rv64-bootstrap-tools-stage1/on-server/busybox
nix hash file --type sha256 --base16 ./result-rv64-bootstrap-tools-stage1/on-server/bootstrap-tools.tar.xz
# ... hashes to fill

nix store add-path ./result-rv64-bootstrap-tools-stage1/on-server/busybox
nix store add-file ./result-rv64-bootstrap-tools-stage1/on-server/bootstrap-tools.tar.xz
```

6. Bootstrap again to check if it works. But not update hashes this time.
```shell
nix build -f pkgs/stdenv/linux/make-bootstrap-tools.nix build --arg localSystem '{ system = "riscv64-linux"; }' -L --keep-going --builders 'ssh://rv64_machine_or_vm riscv64-linux - 4' -j0 -o result-rv64-bootstrap-tools-stage2
```

###### Notes

1. Bootstrap tools are currently using fake URLs. They should be uploaded first and updated to `tarballs.nixos.org` before merging this PR.
2. All intermediate and final store paths are available on <https://riscv64.cachix.org>.
3. Using `binfmt` instead of native riscv64 platform (or QEMU VM) will make few derivations like `coreutils` fail the `checkPhase` because of changed `argv[0]` and emulator behavior. Some tests expects a specific failure like "invalid cwd" from binary, but binfmt qemu fails before it which produces different error message.

4. ~`gcc10` uses PGO and therefore not deterministic (#108475). So bootstrap file hashs may also be not reproducible.~
5. `libatomic` is required for packages linking `pthread` or using `<atomic>` (like `xz`) during the bootstrap. ~But there is another way to avoid producing `libatomic` in bootstrap tools, but re-configure these packages to disable multi-threading before new `gcc` is built (like [what we do](https://github.com/NixOS/nixpkgs/blob/b2a42f017ff3308a4f6394bc949a8acb8e8c8be2/pkgs/stdenv/linux/default.nix#L210) with `perl`)
   I'm not sure which way is preferred.~
   I just added `libatomic` to bootstrap tools.
6. `bootstrap-tools-stage2` differs from `bootstrap-tools-stage1`. The differences are from glibc and gcc (cc1/cc1plus). 
   I believe it's caused by that `-frandom-seed` is based on `$out` (#102251), which is hashed from all drv inputs. Here between two bootstraps, the bootstrap tools should be functionally identical but have different hash, which changes `random-seed` of downstream packages.
   It should not be an issue though.

<details>
<summary><code>diff -r ./result-rv64-bootstrap-tools-stage{1,2}</code></summary>

```
Binary files ./result-rv64-bootstrap-tools-stage1/on-server/bootstrap-tools.tar.xz and ./result-rv64-bootstrap-tools-stage2/on-server/bootstrap-tools.tar.xz differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/bin/make and ./result-rv64-bootstrap-tools-stage2/pack/bin/make differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/crt1.o and ./result-rv64-bootstrap-tools-stage2/pack/lib/crt1.o differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/ld-2.33.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/ld-2.33.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/ld-linux-riscv64-lp64d.so.1 and ./result-rv64-bootstrap-tools-stage2/pack/lib/ld-linux-riscv64-lp64d.so.1 differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libc-2.33.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libc-2.33.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libcrypt-2.33.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libcrypt-2.33.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libcrypt.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libcrypt.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libcrypt.so.1 and ./result-rv64-bootstrap-tools-stage2/pack/lib/libcrypt.so.1 differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libc.so.6 and ./result-rv64-bootstrap-tools-stage2/pack/lib/libc.so.6 differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libdl-2.33.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libdl-2.33.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libdl.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libdl.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libdl.so.2 and ./result-rv64-bootstrap-tools-stage2/pack/lib/libdl.so.2 differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libm-2.33.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libm-2.33.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libmemusage.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libmemusage.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libm.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libm.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libm.so.6 and ./result-rv64-bootstrap-tools-stage2/pack/lib/libm.so.6 differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libnsl-2.33.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libnsl-2.33.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libnsl.so.1 and ./result-rv64-bootstrap-tools-stage2/pack/lib/libnsl.so.1 differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libnss_compat-2.33.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libnss_compat-2.33.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libnss_compat.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libnss_compat.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libnss_compat.so.2 and ./result-rv64-bootstrap-tools-stage2/pack/lib/libnss_compat.so.2 differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libnss_db-2.33.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libnss_db-2.33.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libnss_db.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libnss_db.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libnss_db.so.2 and ./result-rv64-bootstrap-tools-stage2/pack/lib/libnss_db.so.2 differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libnss_dns-2.33.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libnss_dns-2.33.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libnss_dns.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libnss_dns.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libnss_dns.so.2 and ./result-rv64-bootstrap-tools-stage2/pack/lib/libnss_dns.so.2 differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libnss_files-2.33.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libnss_files-2.33.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libnss_files.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libnss_files.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libnss_files.so.2 and ./result-rv64-bootstrap-tools-stage2/pack/lib/libnss_files.so.2 differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libnss_hesiod-2.33.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libnss_hesiod-2.33.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libnss_hesiod.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libnss_hesiod.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libnss_hesiod.so.2 and ./result-rv64-bootstrap-tools-stage2/pack/lib/libnss_hesiod.so.2 differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libpthread-2.33.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libpthread-2.33.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libpthread.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libpthread.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libpthread.so.0 and ./result-rv64-bootstrap-tools-stage2/pack/lib/libpthread.so.0 differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libresolv-2.33.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libresolv-2.33.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libresolv.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libresolv.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libresolv.so.2 and ./result-rv64-bootstrap-tools-stage2/pack/lib/libresolv.so.2 differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/librt-2.33.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/librt-2.33.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/librt.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/librt.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/librt.so.1 and ./result-rv64-bootstrap-tools-stage2/pack/lib/librt.so.1 differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libutil-2.33.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libutil-2.33.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libutil.so and ./result-rv64-bootstrap-tools-stage2/pack/lib/libutil.so differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/lib/libutil.so.1 and ./result-rv64-bootstrap-tools-stage2/pack/lib/libutil.so.1 differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/libexec/gcc/riscv64-unknown-linux-gnu/10.3.0/cc1 and ./result-rv64-bootstrap-tools-stage2/pack/libexec/gcc/riscv64-unknown-linux-gnu/10.3.0/cc1 differ
Binary files ./result-rv64-bootstrap-tools-stage1/pack/libexec/gcc/riscv64-unknown-linux-gnu/10.3.0/cc1plus and ./result-rv64-bootstrap-tools-stage2/pack/libexec/gcc/riscv64-unknown-linux-gnu/10.3.0/cc1plus differ
```

</details>
</details>

oxalica committed:
bintools-wrapper: add dynamicLinker for riscv

oxalica committed:
make-bootstrap-tools: produce libatomic on riscv platform

SuperSandro2000 commented on pkgs/build-support/bintools-wrapper/default.nix:
```diff
@@ -64,6 +64,7 @@ let
     else if targetPlatform.system == "aarch64-linux"  then "${libc_lib}/lib/ld-linux-aarch64.so.1"
     else if targetPlatform.system == "powerpc-linux"  then "${libc_lib}/lib/ld.so.1"
     else if targetPlatform.isMips                     then "${libc_lib}/lib/ld.so.1"
+    else if targetPlatform.system == "riscv64-linux"  then "${libc_lib}/lib/ld-linux-riscv64-lp64d.so.1"
```
Maybe we want to add targetPlatform.isRiscv64 ?

oxalica replied:
@SuperSandro2000 
> Maybe we want to add targetPlatform.isRiscv64 ?

`lp64d` (part of `so` name) is just one of riscv64 ABIs. User-defined riscv64 platform may use other ABIs.

Ericson2314 replied:
Any user defined one would still use `system = "riscv64-linux"` however. Extra ABI info should go inside `targetPlatform.parsed.abi` somehow. See https://github.com/NixOS/nixpkgs/blob/master/lib/systems/parse.nix#L307-L361

oxalica replied:
We only have `isRiscV` instead of  `isRiscV64`. Maybe we could use `"${libc_lib}/lib/ld-linux-riscv*.so.1"` like what we did for ARM?

Ericson2314 replied:
Yeah that should be possible I think.

oxalica commented:
cc @r-burns

zhaofengli commented:
Any progress on this? Would be great if a trusted member could produce an official bootstrap tools tarball on https://tarballs.nixos.org

Ericson2314 commented on pkgs/development/tools/misc/libtool/libtool2.nix:
```diff
@@ -15,6 +15,9 @@ stdenv.mkDerivation rec {
   nativeBuildInputs = [ perl help2man m4 ];
   propagatedBuildInputs = [ m4 ];
 
+  # FIXME: Configure script is too old (2015) to be able to guess RiscV platform.
+  configureFlags = lib.optional stdenv.buildPlatform.isRiscV "--build=${stdenv.buildPlatform.config}";
```
Huh, are you sure this is necessary? Doesn't https://github.com/NixOS/nixpkgs/blob/32c2087987ac9d4f13d80780d97ee6038ffacdc9/pkgs/stdenv/linux/default.nix#L391-L392 and similar logic make this unnecessary?

oxalica committed:
stdenv: patch autoconf script for riscv in stage 2

oxalica commented:
@Ericson2314 
What do you think about issue 2 about `libatomic` from gcc-lib? Currently I simply copy it into bootstrap tools. We can also not touching bootstrap tools, but disable multi-threading pkgs like `xz` before the new `gcc` is built (in `bootstrap-stage4`).

Ericson2314 commented:
@oxalica I don't mind adding to the the bootstrap tools. Long term, I am hoping the bootstrap tools can *just* be the tools, statically linked, and we bootstrap all the run-time libraries up from `libgcc`/`compiler-rt. But that is a big refactor, and will start with the better-behaved LLVM stuff first, so I don't care if we add more one `libatomic.a` in the meantime.

oxalica commented:
Rebased and rebuilt the seed.

zhaofengli commented:
Gentle poke. Now that [a basic NixOS system](https://github.com/NixOS/nixpkgs/issues/101651#issuecomment-852725823) is running on riscv64, I'm really looking forward to seeing this PR merged in the near future. [This script](https://gist.github.com/zhaofengli/275d1a0de14eceba90fd4e399af34b5c) starts a simple qemu-system-riscv64 VM and should help with natively building the bootstrap tools for maintainers without access to RISC-V hardware.

zhaofengli commented:
Hmm, I have a problem. The GCC in stdenv doesn't seem to be able to find spec files:

```
$ echo "int main() {}" > test.c
$ /nix/store/57q4bwd5hz1vpxr4fx7px14qxay9aw00-gcc-wrapper-10.3.0/bin/gcc -fopenmp test.c
gcc: fatal error: cannot read spec file ‘libgomp.spec’: No such file or directory
compilation terminated.
```

`libgomp.spec` is indeed in the `lib` of the wrapped compiler, and I can link libgomp directly myself:

```
$ ls /nix/store/dky7mm6fg9b30cqif055m1h8k91kihjs-gcc-10.3.0/lib
gcc        libasan_preinit.o  libcc1.a   libgomp.spec       libssp.a            libstdc++.a    libsupc++.a
libasan.a  libatomic.a        libgomp.a  libsanitizer.spec  libssp_nonshared.a  libstdc++fs.a  libubsan.a
$ cat /nix/store/dky7mm6fg9b30cqif055m1h8k91kihjs-gcc-10.3.0/lib/libgomp.spec
# This spec file is read by gcc when linking.  It is used to specify the
# standard libraries we need in order to link with libgomp.
*link_gomp: -lgomp %{static: -ldl -lpthread }
$ /nix/store/57q4bwd5hz1vpxr4fx7px14qxay9aw00-gcc-wrapper-10.3.0/bin/gcc -lgomp test.c
$
```

The same commands work fine on an x86-64 host.

zhaofengli commented:
Still kind of confused though there's a solution. Here is the list of paths that gcc on x86-64 searched:

```
/nix/store/54klr10i53jdfgn7322mzgza6wsai0q8-gcc-10.3.0-lib/lib/x86_64-unknown-linux-gnu/10.3.0/libgomp.spec
/nix/store/54klr10i53jdfgn7322mzgza6wsai0q8-gcc-10.3.0-lib/lib/libgomp.spec
/nix/store/sbbifs2ykc05inws26203h0xwcadnf0l-glibc-2.32-46/lib/x86_64-unknown-linux-gnu/10.3.0/libgomp.spec
/nix/store/sbbifs2ykc05inws26203h0xwcadnf0l-glibc-2.32-46/lib/libgomp.spec
/nix/store/35pnk5kwi26m3ph2bc7dxwjnavpzl8cn-gcc-wrapper-10.3.0/bin/x86_64-unknown-linux-gnu/10.3.0/libgomp.spec
/nix/store/35pnk5kwi26m3ph2bc7dxwjnavpzl8cn-gcc-wrapper-10.3.0/bin/libgomp.spec
/nix/store/h3f8rn6wwanph9m3rc1gl0lldbr57w3l-gcc-10.3.0/lib/gcc/x86_64-unknown-linux-gnu/10.3.0/libgomp.spec
/nix/store/h3f8rn6wwanph9m3rc1gl0lldbr57w3l-gcc-10.3.0/lib/gcc/x86_64-unknown-linux-gnu/10.3.0/../../../../x86_64-unknown-linux-gnu/lib/x86_64-unknown-linux-gnu/10.3.0/libgomp.spec
/nix/store/h3f8rn6wwanph9m3rc1gl0lldbr57w3l-gcc-10.3.0/lib/gcc/x86_64-unknown-linux-gnu/10.3.0/../../../../x86_64-unknown-linux-gnu/lib/../lib64/libgomp.spec
/nix/store/h3f8rn6wwanph9m3rc1gl0lldbr57w3l-gcc-10.3.0/lib/gcc/x86_64-unknown-linux-gnu/10.3.0/../../../x86_64-unknown-linux-gnu/10.3.0/libgomp.spec
/nix/store/h3f8rn6wwanph9m3rc1gl0lldbr57w3l-gcc-10.3.0/lib/gcc/x86_64-unknown-linux-gnu/10.3.0/../../../../lib64/libgomp.spec
  -> hurray, we symlink $out/lib64 to $out/lib - success
```

The same applies on aarch64. However, on riscv64 it tries the following before giving up:

```
/nix/store/zw9qnha64sygasja7am4fhmlx5jx4yhb-gcc-10.3.0-lib/lib/riscv64-unknown-linux-gnu/10.3.0/libgomp.spec
/nix/store/zw9qnha64sygasja7am4fhmlx5jx4yhb-gcc-10.3.0-lib/lib/libgomp.spec
/nix/store/m1jbr8vqnpgnpjfdzx13da897pd3aa6f-glibc-2.32-46/lib/riscv64-unknown-linux-gnu/10.3.0/libgomp.spec
/nix/store/m1jbr8vqnpgnpjfdzx13da897pd3aa6f-glibc-2.32-46/lib/libgomp.spec
/nix/store/57q4bwd5hz1vpxr4fx7px14qxay9aw00-gcc-wrapper-10.3.0/bin/riscv64-unknown-linux-gnu/10.3.0/libgomp.spec
/nix/store/57q4bwd5hz1vpxr4fx7px14qxay9aw00-gcc-wrapper-10.3.0/bin/libgomp.spec
/nix/store/dky7mm6fg9b30cqif055m1h8k91kihjs-gcc-10.3.0/lib/gcc/riscv64-unknown-linux-gnu/10.3.0/libgomp.spec
/nix/store/dky7mm6fg9b30cqif055m1h8k91kihjs-gcc-10.3.0/lib/gcc/riscv64-unknown-linux-gnu/10.3.0/../../../../riscv64-unknown-linux-gnu/lib/riscv64-unknown-linux-gnu/10.3.0/libgomp.spec
/nix/store/dky7mm6fg9b30cqif055m1h8k91kihjs-gcc-10.3.0/lib/gcc/riscv64-unknown-linux-gnu/10.3.0/../../../../riscv64-unknown-linux-gnu/lib/libgomp.spec
/lib64/lp64d/riscv64-unknown-linux-gnu/10.3.0/libgomp.spec
/lib64/lp64d/libgomp.spec
/usr/lib64/lp64d/riscv64-unknown-linux-gnu/10.3.0/libgomp.spec
/usr/lib64/lp64d/libgomp.spec
/lib/riscv64-unknown-linux-gnu/10.3.0/libgomp.spec
/lib/libgomp.spec
/usr/lib/riscv64-unknown-linux-gnu/10.3.0/libgomp.spec
/usr/lib/libgomp.spec
```

At first glance we may want to link `$out/[triplet]/lib` to `$out/lib` as well, but I feel uneasy adopting a solution without knowing why. `gcc/config/riscv/linux.h` has [these lines](https://github.com/gcc-mirror/gcc/blob/releases/gcc-10.3.0/gcc/config/riscv/linux.h#L72-L76) that may explain the `lib/lp64d` paths, but not the others. This seems to be handled in [`set_up_specs`](https://github.com/gcc-mirror/gcc/blob/releases/gcc-10.3.0/gcc/gcc.c#L7570) which follows the startfile prefixes, but I'm pretty unfamiliar with the code.

zhaofengli commented:
I've tested [a naive fix](https://github.com/zhaofengli/nixpkgs/commit/a624f7923f67eee4fc576552074d18e3b17738e1) which does make `-fopenmp` work in the final native toolchain. Would like to hear whether there is a more proper fix to the issue.

cc @Ericson2314 @oxalica

vcunat commented:
> I'm wait for the coreutils 9.0 in staging-next #141684. Hope we can have it for linux.

It is in master now, though there were complications because of a ZFS bug.

oxalica commented:
Rebased and removed commits containing binary hashes.

@r-burns
I'd request a Hydra job to build cross compiled bootstrap tools on this PR, basically,
`nix build -f pkgs/stdenv/linux/make-bootstrap-tools-cross.nix riscv64.build`
And upload two result binaries as other bootstrap tools did.

vcunat commented:
So we simply merge this and it will appear there, right? https://hydra.nixos.org/job/nixpkgs/cross-trunk/bootstrapTools.riscv64.bootstrapTools

(And continue in followup PRs.)

vcunat commented:
I expect the builds to finish within a few hours: https://hydra.nixos.org/eval/1726433?filter=riscv64

oxalica commented:
It's done building and I can match result hashes with my own build.

But how can we reference these tools by `tarball.nixos.org` like others?
https://github.com/NixOS/nixpkgs/blob/33da2dcc496668974369e15667db2c47e2dde6b7/pkgs/stdenv/linux/bootstrap-files/aarch64.nix#L3

Do we need someone manually upload build results of [`busybox`](https://hydra.nixos.org/build/159891432) and [`bootstrap-tools.tar.xz`](https://hydra.nixos.org/build/159891436)?

vcunat commented:
For the commit in URL it's customary to use the one that generated the bootstrap, i.e. `9bd3cf0063b80428bd85a286205adab4b6ffcbd6` here.  The rest is probably obvious?

I'd suggest preparing the PR and test that the bootstrap works locally (if you prefetch the files, nix won't attempt download).  Then ping someone who can upload it manually; I'd suggest `@grahamc`.  After upload, perhaps re-test and then merge.

oxalica commented:
I've opened the following PR #147942.

> I'd suggest preparing the PR and test that the bootstrap works locally (if you prefetch the files, nix won't attempt download).

I had tested building `stdenv` with my previous built bootstrap files. And Hydra reproduced my bootstrap files. So I could simply change URLs of FODs, which doesn't change any downstream derivations.
